### PR TITLE
MXNet: support Gluon Trainer API

### PR DESCRIPTION
**Pull request by yuxihu**
This PR makes the following changes:
1. support broadcast parameter which type is ParameterDict
2. update imagenet example to provide an option to train with Gluon Trainer API

### CODE REVIEW

**Comment by yuxihu** [comment]
@apeforest please review

File: README.md [hunk 1]
```diff
@@ -234,9 +237,17 @@ model = ...
...
 
 # Add Horovod Distributed Optimizer
-opt = mx.optimizer.create('sgd', sym=model, **optimizer_params)
+opt = mx.optimizer.create('sgd', **optimizer_params)
 opt = hvd.DistributedOptimizer(opt)
 
+# Initialize parameters
+initializer = mx.init.Xavier(rnd_type='gaussian', factor_type="in",
+                             magnitude=2)
+mod.bind(data_shapes=train_data.provide_data,
```
Comment (apeforest) [inline README.md]: model.bind

File: README.md [hunk 2]
```diff
@@ -234,9 +237,17 @@ model = ...
...
+         label_shapes=train_data.provide_label)
+mod.init_params(initializer)
```
Comment (apeforest) [inline README.md]: model.init_params

File: README.md [hunk 3]
```diff
@@ -216,6 +216,9 @@ See full training [MNIST](examples/mxnet_mnist.py) and [ImageNet](examples/mxnet
 
 **Note**: we recommend users to build MXNet from source following this [guide](https://mxnet.incubator.apache.org/install/build_from_source.html) when running Horovod with MXNet on a Linux OS with GCC version 5.X and above. The MXNet shared library distributed through MXNet pip package is currently built using GCC 4.8.4. If we build and install Horovod on a Linux OS with GCC 5.X+ with MXNet pip package, we will hit segmentation fault due to std::function definition change from GCC [4.X](https://github.com/gcc-mirror/gcc/blob/gcc-4_8_4-release/libstdc++-v3/include/std/functional#L2069) to GCC [5.X](https://github.com/gcc-mirror/gcc/blob/gcc-5_4_0-release/libstdc++-v3/include/std/functional#L1854).
 
+There are two ways to train a model using MXNet: Module Symbolic API and Gluon Trainer API. Here we provide the building block for each set of API to train a model using MXNet with Horovod.
```
Comment (apeforest) [inline README.md]: Maybe we can add hyperlink http://mxnet.incubator.apache.org/api/python/gluon/gluon.html and http://mxnet.incubator.apache.org/api/python/module/module.html for Gluon and Symbolic APIs respectively?

File: README.md [hunk 4]
```diff
@@ -247,9 +258,59 @@ model.set_params(arg_params=arg_params, aux_params=aux_params)
...
+# Initialize parameters
+model.initialize(initializer, ctx=context)
+
+# Fetch and broadcast parameters
+params = model.collect_params()
+if params is not None:
+    hvd.broadcast_parameters(params, root_rank=0)
+
+# Create trainer and loss function
+trainer = gluon.Trainer(params, opt, kvstore=None)
```
Comment (apeforest) [inline README.md]: Is kvstore=None required? Can we leave it as default?

File: README.md [hunk 5]
```diff
@@ -247,9 +258,59 @@ model.set_params(arg_params=arg_params, aux_params=aux_params)
...
+loss_fn = gluon.loss.SoftmaxCrossEntropyLoss()
```
Comment (apeforest) [inline README.md]: Since this is just a high-level pseudo code, maybe we just say loss = ... ?

Comment (yuxihu) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (yuxihu) [reply README.md]: Done

File: README.md (hunk 2 above)
Comment (yuxihu) [reply README.md]: Done

File: README.md (hunk 1 above)
Comment (yuxihu) [reply README.md]: Done

File: examples/mxnet_imagenet_resnet50.py [hunk 6]
```diff
@@ -73,16 +72,23 @@
                     each bottleneck to 0 (default: False)')
 parser.add_argument('--model', type=str, default='resnet50_v1',
                     help='type of model to use. see vision_model for options.')
+parser.add_argument('--mode', type=str, default='symbolic',
+                    help='mode in which to train the model. options are \
+                    symbolic, hybrid (default: symbolic)')
```
Comment (apeforest) [inline examples/mxnet_imagenet_resnet50.py]: maybe use the name gluon instead of hybrid?

File: README.md (hunk 4 above)
Comment (yuxihu) [reply README.md]: I think it is not required but better to have. We already wrapped the opt with DistributedOptimizer so the kvstore code path will not be executed during training. However, explicitly setting it to None will skip some unnecessary checking internally. And it is also informative that we do not use kvstore with Horovod.

File: examples/mxnet_imagenet_resnet50.py (hunk 6 above)
Comment (yuxihu) [reply examples/mxnet_imagenet_resnet50.py]: Done

**Comment by yuxihu** [comment]
@alsrgv This PR is ready to go. Please help review and merge. Thanks.

**Comment by alsrgv** [comment]
Thanks!
